cmd/harmony/config/config.go

```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
 	}
 }
 
-func getDefaultSyncConfig(nt nodeconfig.NetworkType) syncConfig {
+func GetDefaultSyncConfig(nt nodeconfig.NetworkType) SyncConfig {
 	switch nt {
 	case nodeconfig.Mainnet:
-		return defaultMainnetSyncConfig
+		return DefaultMainnetSyncConfig
 	case nodeconfig.Testnet:
-		return defaultTestNetSyncConfig
+		return DefaultTestNetSyncConfig
 	case nodeconfig.Localnet:
 		return defaultLocalNetSyncConfig
 	default:
 		return defaultElseSyncConfig
 	}
 }
 
-var dumpConfigCmd = &cobra.Command{
-	Use:   "dumpconfig [config_file]",
-	Short: "dump the config file for harmony binary configurations",
-	Long:  "dump the config file for harmony binary configurations",
-	Args:  cobra.MinimumNArgs(1),
-	Run: func(cmd *cobra.Command, args []string) {
-		nt := getNetworkType(cmd)
-		config := getDefaultHmyConfigCopy(nt)
-
-		if err := writeHarmonyConfigToFile(config, args[0]); err != nil {
-			fmt.Println(err)
-			os.Exit(128)
-		}
-	},
-}
-
-func registerDumpConfigFlags() error {
-	return cli.RegisterFlags(dumpConfigCmd, []cli.Flag{networkTypeFlag})
+func GetDefaultDNSSyncConfig(nt nodeconfig.NetworkType) DNSSyncConfig {
+	config := DNSSyncConfig{
+		LegacySyncing: false,
+		DNSZone:       nodeconfig.GetDefaultDNSZone(nt),
+		DNSPort:       nodeconfig.GetDefaultDNSPort(nt),
+		LegacyServer:  true,
+		LegacyClient:  true,
+	}
+	switch nt {
+	case nodeconfig.Mainnet:
+	case nodeconfig.Testnet:
+	case nodeconfig.Localnet:
+	default:
+		config.LegacyClient = false
+	}
+	return config
 }
 
-func loadHarmonyConfig(file string) (harmonyConfig, error) {
+func LoadHarmonyConfig(file string) (HarmonyConfig, error) {
 	b, err := ioutil.ReadFile(file)
 	if err != nil {
-		return harmonyConfig{}, err
+		return HarmonyConfig{}, err
 	}
-
-	var config harmonyConfig
-	if err := toml.Unmarshal(b, &config); err != nil {
-		return harmonyConfig{}, err
+	var header struct {
+		Version string
+	}
+	if err := toml.Unmarshal(b, &header); err != nil {
+		return HarmonyConfig{}, fmt.Errorf("reading config version: %w", err)
+	}
+	var config HarmonyConfig
+	// lexicographicaly compare
+	if header.Version <= v1TOMLConfigVersion {
+		config, err = loadHarmonyConfigV1(b)
+	} else {
+		err = toml.Unmarshal(b, &config)
```

Comment:
Refactor this to `loadHarmonyConfigV2` would increase readability IMO.

Reply:
without renaming the struct itself I don't see a point in renaming/refactoring this one into standalone func. the whole idea is to have the latest and greatest HarmonyConfig and related functions public and exported, and other legacy structs and code to work with them private and hidden from other code. 

this way the action list for adding a new (backward incompatible) config version would be:
* copy current struct into private struct with version name in it and put it into a standalone file, say v2.go
* change HarmonyConfig to represent the newest version of config
* add a private loadV2Config() function to v2.go to convert old to new, and call it in the LoadHarmonyConfig publicly exported function.

if you have another plan in mind, could you please elaborate how do you see an action list for changing config to the new incompatible version would be with it?

Reply:
> * copy current struct into private struct with version name in it and put it into a standalone file, say v2.go
> * change HarmonyConfig to represent the newest version of config
> * add a private loadV2Config() function to v2.go to convert old to new, and call it in the LoadHarmonyConfig publicly exported function.

Yes, you nailed it! 

And here I say refactor is just to refactor this one line `toml.Unmarshal` to a function to increase readability